### ISSUE TITLE
Eager compute template digests in production

### ISSUE THREAD

**Issue by dhh** [rails/rails#23901]
Computing digests for templates at run-time risks burdening individual requests with that work, which may well take tens of milliseconds. We should have an option, and it should probably be default, to preload all template digests at server boot.

**Comment by dhh** [comment]
@kaspth Thoughts?

**Comment by saiqulhaq** [comment]
could this be related to https://github.com/railsgsoc/ideas/wiki/2016-Ideas#eager-load-action-view-templates ?

**Comment by dhh** [comment]
Ah yes, same task.

**Comment by saiqulhaq** [comment]
ok i though so
please let rails GSOC students do this task 
I think many students will have interest to do this task, including me 

hi @tenderlove , you will be the mentor for this rails GSOC idea, right? how do you think?

**Comment by kaspth** [comment]
Yeah, reading this it reminded me of the eager loaded templates thing @tenderlove had been looking into. It's not a given this will be rolled up into that project, it might be too big. If so, we could look into it after the that GSoC project is through, since it would likely be much easier to implement after it.

But yeah, 👍 on the whole concept.

**Comment by st0012** [comment]
Hello guys, I am currently working on a gem called [zubot](https://github.com/st0012/zubot). Which can now do this job for Rails apps.
I even solved the local variable issue for precompiling partials, which might be the biggest obstacle for implementing this feature (but my approach might have some drawbacks).
It's not production ready yet and some places uses monkey-patching or very naive implementation, but I think it can achieve the goal of this issue. Hope you guys can give me some feedback. I have working on this subject for half years and I am happy to implement it 😁.
cc @kaspth @tenderlove

**Comment by kaspth** [comment]
@st0012 hey Stan, thanks for chiming in! You might also want to look into the original PR from some Open Academy students that tried to work on this first: https://github.com/rails/rails/pull/19272.

How does your solution solve the template locals affecting the cache part?

**Comment by st0012** [comment]
Thanks for the reply @kaspth 😄 

Actually I wrote my solution at my [gem's readme](https://github.com/st0012/zubot#the-solution).

But in short my approach is to store local assigns in the compiled method like
```ruby
  source = <<-end_src
    def #{method_name}(local_assigns, output_buffer)
      # Assigns local to instance variable
      @local_assigns = local_assigns

       # Actually we might not need the locals_code here anymore
      _old_virtual_path, @virtual_path = @virtual_path, #{@virtual_path.inspect};_old_output_buffer = @output_buffer;#{locals_code};#{code}
    ensure
      @virtual_path, @output_buffer = _old_virtual_path, _old_output_buffer
    end
  end_src
```

And use `method_missing` to access them:

```ruby
module ActionView
  module MethodMissing
    def method_missing(name, *args, &block)
      if @local_assigns && local = @local_assigns[name]
        local
      else
        super
      end
    end
  end

  class Base
    include MethodMissing
  end
end
```

However, this approach will conflict with other gems that also use `method_missing` in `ActionView::Base`, like `kaminari` and I'm still trying to fix it.
And the other concern is that I don't know if using `method_missing` would have huge impact on performance 😕. Maybe we can find a way to catch the result after first `method_missing` was called.

**Comment by matthewd** [comment]
Two related-but-distinct patterns come to mind:

```
<%
unless defined?(foo)
  foo = 'hello world'
end
%>
<h1><%= foo %></h1>
```

---

```
<% foo = 'hello' unless bar %>
<h1><%= foo %></h1>
```

**Comment by st0012** [comment]
@matthewd New approach:
```ruby
source = <<-end_src
        def #{method_name}(local_assigns, output_buffer)
          local_assigns.each do |key, value|
            self.define_singleton_method(key) { value } unless methods.include?(key)
          end
          _old_virtual_path, @virtual_path = @virtual_path, #{@virtual_path.inspect};_old_output_buffer = @output_buffer;#{locals_code};#{code}
        ensure
          @virtual_path, @output_buffer = _old_virtual_path, _old_output_buffer
        end
      end_src
```
I thinks we can define the local binding method on the fly. Do you think this could work?

**Comment by st0012** [comment]
@matthewd @kaspth  My new approach is to create binding method on the fly:

```ruby
  source = <<-end_src
    def #{method_name}(local_assigns, output_buffer)
      @local_assigns = local_assigns

      local_assigns.each_key do |key|
        unless methods.include?(key)
          source = <<-inner_source
            def \#{key}
              @local_assigns[:\#{key}]
            end
          inner_source
          self.instance_eval(source)
        end
      end
      _old_virtual_path, @virtual_path = @virtual_path, #{@virtual_path.inspect};_old_output_buffer = @output_buffer;#{locals_code};#{code}
    ensure
      @virtual_path, @output_buffer = _old_virtual_path, _old_output_buffer
    end
  end_src
```

When the source method was called, we store given local_assigns (say `{ title: "Hello" }`) into the instance variable. And then create a binding method `title` for accessing the local_assigns if there's no `title` method. So Rails can use the `title` method to access the locals.

I also added a test case for this in [here](https://github.com/st0012/zubot/blob/master/spec/template_precompiler/erb_spec.rb#L51) to make sure I am not a fixed value like `"Hello"`.

**Comment by matthewd** [comment]
@st0012 I don't believe that addresses either of the partials I showed, if that was your intention.

**Comment by st0012** [comment]
Hi @matthewd , I think I figured out how to deal with the pattern you mentioned but it's a little too complex for me to explain my approach in English 😓.

So if you have some time I wish you can take a look at my test cases for this situation.

## Pattern1

[Test case](https://github.com/st0012/zubot/blob/master/spec/template_precompiler/erb_spec.rb#L90)
```ruby
      it "returns \"Test\" instead of \"hello world\" in pattern 1" do
        subject.compile_template(implicit_file_path("posts/local_check/pattern1.html.erb"), resolver)
        subject.compile_template(implicit_file_path("posts/local_check/_pattern1_partial.html.erb"), resolver)

        expect_any_instance_of(ActionView::Template).not_to receive(:compile)

        result = view.render(template: "posts/local_check/pattern1")

        expect(result).to eq("<h1>Test</h1>\n\n")
      end
```

[Template](https://github.com/st0012/zubot/blob/master/spec/fixtures/templates/posts/local_check/pattern1.html.erb)

```erb
<%= render "posts/local_check/pattern1_partial", foo: "Test" %>
```

[Partial](https://github.com/st0012/zubot/blob/master/spec/fixtures/templates/posts/local_check/_pattern1_partial.html.erb)

```erb

<%
unless defined?(foo)
  foo = 'hello world'
end
%>
<h1><%= foo %></h1>
```


## Pattern2

[Test case](https://github.com/st0012/zubot/blob/master/spec/template_precompiler/erb_spec.rb#L100)

```ruby
      it "returns \"Test\" instead of \"hello world\" in pattern 2" do
        subject.compile_template(implicit_file_path("posts/local_check/pattern2.html.erb"), resolver)
        subject.compile_template(implicit_file_path("posts/local_check/_pattern2_partial.html.erb"), resolver)

        expect_any_instance_of(ActionView::Template).not_to receive(:compile)

        result = view.render(template: "posts/local_check/pattern2")

        expect(result).to eq("<h1>Test</h1>\n\n")
      end
```

[Template](https://github.com/st0012/zubot/blob/master/spec/fixtures/templates/posts/local_check/pattern2.html.erb)

```erb
<%= render "posts/local_check/pattern2_partial", foo: "Test", bar: true %>
```

[Partial](https://github.com/st0012/zubot/blob/master/spec/fixtures/templates/posts/local_check/_pattern2_partial.html.erb)

```erb
<% foo = 'hello' unless bar %>
<h1><%= foo %></h1>
```


And here's the [implementation](https://github.com/st0012/zubot/blob/master/lib/zubot/actionview/template.rb#L20)

Thank you for your advice, I didn't think about this kind of situations before.

**Comment by matthewd** [comment]
That's not really precompiling; that's getting back to the same cache-per-locals behaviour we already have.

---

Another one:

```ruby
module MyHelpers
  def foo
    'helper'
  end
end
```

```
<%= render "posts/local_check/pattern3_partial", foo: "local" %>
<%= render "posts/local_check/pattern3_partial" %>
```

```
<h1><%= foo %></h1>
```

Or the same thing, without a helper method:

```
<%= render "posts/local_check/pattern4_partial", foo: "Test" %>
<%= render "posts/local_check/pattern4_partial" %>
```

```
<h1><%= defined?(foo) || 'hello' %></h1>
```

**Comment by st0012** [comment]
@matthewd I think it's the way I wrote tests made you think that it's not precompile, so I updated my [test suites](https://github.com/st0012/zubot/blob/master/spec/template_precompiler/erb_spec.rb). I only run precompile once at [here](https://github.com/st0012/zubot/blob/master/spec/template_precompiler/erb_spec.rb#L7), and this should be the only compilation it requires to finish all test cases below.

**Comment by matthewd** [comment]
It's using eval to redefine the method at runtime.

**Comment by st0012** [comment]
Yes, it's not really `precompile` since the view method's content might still be changed. But
1. This only happens when it's rendering a partial with local variables
2. This still can prevent finding template object multiple times in different processes and skip other work we need to do in the compilation process.

**Comment by st0012** [comment]
@matthewd After several experiments, this is the best way I can come up with 😞. Do you have any possible solution on this issue?

**Comment by kaspth** [comment]
@st0012 Let's revisit this after Rails 5.1 is out 😊
